File: 0695-max-area-of-island/0695-max-area-of-island.cpp
class Solution {
public:
    
    vector<vector<int>>directions{{1,0},{-1,0},{0,1},{0,-1}};
    
    int dfs(vector<vector<int>>&grid,int i,int j){
            if(i<0 or i>=grid.size() or j<0 or j>=grid[0].size() or grid[i][j]==0){
                        return 0;
            }      
            int cnt=1;
            grid[i][j]=0;
            for(int k=0;k<directions.size();k++){
                    int x=i+directions[k][0];
                    int y=j+directions[k][1];
                    cnt+=dfs(grid,x,y);
            }
            return cnt;
    }
    
    
    
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int max_cnt=0;
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++){
                    if(grid[i][j]==1){
                            // cout<<dfs(grid,i,j)<<endl;
                            int cnt=dfs(grid,i,j);
                            //cout<<cnt<<endl;
                            max_cnt=max(max_cnt,cnt);
                    }
            }
        }
        return max_cnt;
    }
};

File: 0695-max-area-of-island/0695-max-area-of-island.py
class Solution:
    
    
    def __init__(self):
        
        self.directions=[[1,0],[0,1],[-1,0],[0,-1]]
    
    
    def helper(self,grid,i,j):
        
        if i<0 or i>=len(grid) or j<0 or j>=len(grid[0]) or grid[i][j]==0:
                return 0
            
        grid[i][j]=0
        cnt=1
        for item in self.directions:
                x=item[0]+i
                y=item[1]+j
                cnt+=self.helper(grid,x,y)
         
        return cnt
    
    
    
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
    
    
        max_cnt=0
        for i,row in enumerate(grid):
            for j,item in enumerate(row):
                
                if item==1:
                    
                    val=self.helper(grid,i,j)
                    max_cnt=max(max_cnt,val)    
        
                
        return max_cnt
        
        